### ISSUE TITLE
1.0 Punch List

### ISSUE THREAD

**Issue by smarterclayton** [openshift/origin#2628]
Items that must be complete prior to 1.0.  Items on this list should be expanded to issues and then self-assigned and linked to this issue.  In rough priority order.

* [x] rename `osc` to `oc` and `osadm` to `oadm` (https://github.com/openshift/origin/pull/2900)
* [x] enable registry authentication (Andy)
* [x] add descriptions to all v1 API resource fields across all APIs (#2632)
* [x] generate REST API docs from swagger in a markdown/asciidoc/we/e format for use in the public docs (https://github.com/openshift/openshift-docs/pull/434)
* [x] add an 'oc types' command that describes the core types used by OpenShift (services, deployment configs, replication controllers, pods, routes, and volumes) (#2914)
* [x] ensure that we are using `util.HandleError()` instead of `glog.Errorf()` when code eats an error. (#2811)
* [x] image promotion CLI command (be able to take an image or tag and add it somewhere else) (#2861)
* [x] deployments should have a canary mode by default for rolling update - detect readiness failure of first new pod before continuing deployment, otherwise fail (#2637)
* [x] prevent privileged/uid-0/free-uid pods from running on the cluster by default
* [x] create an add volume command that adds things to a pod template (like osc env, but for volumes) (#2643)
* [x] simple CLI command for re-importing an image stream from docker hub (clear annotation and display result) (#2769)
* [x] group commands on `osc help` and `osadm` help by their primary use like `git` and `rhc`, organized for most important commands for a new user first, with advanced commands later (https://github.com/openshift/origin/pull/2897)
* [x] new-build command that creates a new build (#2695)
* [x] expose more configuration options for Kubernetes performance tunables on the master (#2718)
* [x] Implement drain/fill plug for the controllers so that they can be left in a paused, not started state but started via an API endpoint. (https://github.com/openshift/origin/pull/2837)
* [x] command on the CLI to export a resource so it can be passed to create (filter off fields that can only be set once) (#2910)
* [x] ensure server components are tunable by accepting Kubernetes arguments in config (#2893, #2908, #2912)

Post release:
* [x] add a REST API tutorial example
* [x] clean up the core "getting started" flow with real examples

Deferred to 3.0.1:
* [x] recreate should drain old pods before creating new ones (#2638, #2639), and rolling-update should have a tunable for the % of pods to bring up at once (#2828)
* [x] adding volume emptydirs in new-app flows (#2929)
* [x] warn about emptydir pods not being persistent (#3049)
* [x] new-app needs search/list functionality (#2672)
* [x] sweep logging across the project, make CLI/build/deployment logging cleaner, make server logs more obvious, make sure we are not excessively logging

**Comment by ncdc** [comment]
Enable registry authentication?

**Comment by smarterclayton** [comment]
Run-ats:

* Support a global setting (defaults to true for dev, false for ansible) that binds EmptyDir volumes to PVCs - enables devs to always use PVCs and see useful value
* Images should include metadata about which volumes need to survive restart - new-app and UI should take that into account
* All templates should have EmptyDir volumes and that metadata

**Comment by pmorie** [comment]
@smarterclayton 

> warn about emptydir pods not being persistent

What's the need here?

**Comment by smarterclayton** [comment]
If we automatically give user emptydirs (or even just elect not to bind their disks) then any deployment action will destroy their data.  It's important to establish that expectation up front, until we have ootb persistent storage.

**Comment by pmorie** [comment]
@smarterclayton: yep, that's what i figured, but i wondered how you
expected the warning to be communicated -- docs? something in code?

**Comment by smarterclayton** [comment]
new-app code

----- Original Message -----

**Comment by mrunalp** [comment]
@smarterclayton What's the plan for preventing uid 0? Do we need docker/libcontainer support?

**Comment by smarterclayton** [comment]
Paul was going to inspect the image pre execution and reject any without a numeric non zero uid. Ultimately this is more of a stopgap until user namespaces (we have ranged allocated already), but even then we might want to keep the mode around.  I think docker changes would probably be too complex - just pushing forward with user namespaces is probably the best approach.

**Comment by mrunalp** [comment]
@smarterclayton okay, that makes sense. Thanks.

**Comment by markturansky** [comment]
> Support a global setting (defaults to true for dev, false for ansible) that binds EmptyDir volumes to PVCs - enables devs to always use PVCs and see useful value

I'd like to understand these requirements better.

Does that mean when a pod author specifically asks for an EmptyDir to instead a) create a claim in the namespace that would match an available volume and b) replace the EmptyDir with the PVC? Does the bound claim in the namespace ever get deleted?  I can see this as a leak for developer playthings that they throw away but then don't know to purposefully delete their claim.  

I assume that no PVs available means use an EmptyDir as usual.  If this is on by default in dev, do you expect to have clusters already provisioned for this to work?

**Comment by smarterclayton** [comment]
It's out of scope now, will come back and discuss later.

The primary use case is to enable EVERYONE to build templates and tools that work with PVC, and in developer / poc / test environments to expose the PVC tool chain even without getting real persistent storage.  It would be an admin level config flag rather than an explicit choice for end users.

----- Original Message -----

**Comment by smarterclayton** [comment]
Punch list is closed for 1.0.  Remaining items will be in 3.0.1 / 1.1.0
